### PR TITLE
chat: add history functionality. can optionally request backlog.

### LINKED ISSUE #1858: chat: feature request: load backlog when joining a chat
Please do this, its sucks to not have this.

### PR DESCRIPTION
**Pull request by loganallenc**
Upon chat creation, you can specify whether subscribers to the chat are allowed to ask for portions of the chat message backlog. If you specify that they cannot, then they cannot.
If you specify that they can, then they can optionally request a portion of the backlog starting from an index specified by the last path segment. For instance: /~zod/urbit-help/20 would fetch every message from index 20 forward.

I have added a %messages event to batch messages to prevent frequent pump-locks upon high message bandwidth, and am batching messages within backlog requests (500 per batch).

The joiner can request whether they want to receive history for a channel or not. This allows a user to quickly join a channel regardless of whether the channel has a long backlog or not.

Fixes #1858 and #1875.

PS: Joe, I've fixed your issue with dropped messages by implementing the start index in backlog request. I've tested this with repeated pump-locks, and no messages are lost.

### CODE REVIEW

File: pkg/arvo/app/chat-cli.hoon [hunk 1]
```diff
@@ -363,13 +365,20 @@
         ;~  (glue ace)
           (tag %create)
           security
-          ;~(plug path (punt ;~(pfix ace glyph)))
+          path
+          (punt ;~(pfix ace glyph))
+          (punt ;~(pfix ace ;~(pose (cold %| bar) (cold %& pad))))
```
Comment (Fang-) [inline pkg/arvo/app/chat-cli.hoon]: Pretty sure this won't parse correctly, demanding at least two spaces after the `path` (you're punting, but the `glue`'s still there!). You want:

```hoon
          ;~  plug
            path
            (punt ;~(pfix ace glyph))
            (punt ;~(pfix ace ;~(pose (cold %| (jest 'n')) (cold %& (jest 'y')))))
          ==
```

Untested, but you get the idea. Note also the `y` and `n` instead of `& |` to avoid ambiguity with the glyph parsing. (Talk had a parser-primitive for that, should probably bring it back.)

File: pkg/arvo/app/chat-cli.hoon [hunk 2]
```diff
@@ -363,13 +365,20 @@
...
         ;~((glue ace) (tag %delete) path)
         ;~((glue ace) (tag %invite) rw path ships)
         ;~((glue ace) (tag %banish) rw path ships)
       ::
-        ;~((glue ace) (tag %join) ;~(plug targ (punt ;~(pfix ace glyph))))
+        ;~  (glue ace)
+          (tag %join)
+          targ
+          (punt ;~(pfix ace glyph))
+          (punt ;~(pfix ace ;~(pose (cold %| bar) (cold %& pad))))
```
Comment (Fang-) [inline pkg/arvo/app/chat-cli.hoon]: Same as above.

I tried my hand at making an `(cmd-with-args %cmd ~[required] ~[optional])` parser building function but never could get it to compile. Should probably do another attempt, these are getting out of hand.

File: pkg/arvo/app/chat-cli.hoon [hunk 3]
```diff
@@ -613,18 +622,22 @@
...
-      ::  write
-      ?-  security
-        ?(%channel %mailbox)  ~
-        ?(%village %journal)  [our-self ~ ~]
+          ::  write
+          ?-  security
+            ?(%channel %mailbox)  ~
+            ?(%village %journal)  [our-self ~ ~]
+          ==
+          ?~(allow-history %.y u.allow-history)
```
Comment (Fang-) [inline pkg/arvo/app/chat-cli.hoon]: `+fall` exists. But if we're gonna do defaulting to constants, probably do that in the parser instead of in the logic. Same for `;join`.

File: pkg/arvo/sur/rw-security.hoon [hunk 4]
```diff
@@ -0,0 +1,8 @@
+|%
++$  rw-security
```
Comment (Fang-) [inline pkg/arvo/sur/rw-security.hoon]: Can we please not do unrelated renaming/restructuring in the same commits as feature changes?

File: pkg/interface/chat/src/js/components/lib/chat-input.js [hunk 5]
```diff
@@ -25,20 +25,22 @@ export class ChatInput extends Component {
...
-          Date.now(),
-          {
-            text: `${Date.now()}`
-          }
-        );
-      }
-      setTimeout(closure, 1000);
-    };
-    setTimeout(closure, 2000);*/
+    //let closure = () => {
```
Comment (Fang-) [inline pkg/interface/chat/src/js/components/lib/chat-input.js]: lol @ these changes

File: pkg/arvo/sur/chat-store.hoon [hunk 6]
```diff
@@ -30,16 +30,24 @@
...
   $%  [%create =ship =path]         ::  %create: create a mailbox at ~ship/path
       [%delete =path]               ::  %delete: delete a mailbox at path
       [%message =path =envelope]    ::  %message: append a message to mailbox
       [%read =path]                 ::  %read: set mailbox to read
   ==
 ::
++$  chat-action
+  $%  ::  %messages: append a list of messages to mailbox
+      ::
+      [%messages =path envelopes=(list envelope)]
```
Comment (Fang-) [inline pkg/arvo/sur/chat-store.hoon]: Shouldn't `%message` just be deprecated in favor of this?

Comment (Fang-) [reply pkg/arvo/sur/chat-store.hoon]: Also updating types that go over the air has historically given problems when ships on latest and latest-1 tried to communicate. Did you test this case on a fakenet, or are these changes going to be breached in?

File: pkg/arvo/app/chat-hook.hoon [hunk 7]
```diff
@@ -33,42 +34,62 @@
...
-|_  [bol=bowl:gall state-one]
+|_  [bol=bowl:gall state-two]
 ::
 ++  this  .
 ::
 ++  prep
-  |=  old=(unit state-both)
-  ^-  (quip move _this)
+  |=  old=(unit versioned-state)
+  |^  ^-  (quip move _this)
```
Comment (Fang-) [inline pkg/arvo/app/chat-hook.hoon]: What's this about?

File: pkg/arvo/app/chat-hook.hoon [hunk 8]
```diff
@@ -157,14 +181,61 @@
...
   ?~  box  !!
   :_  this
-  [ost.bol %diff %chat-update [%create (slav %p i.pax) pax]]~
+  :-  [ost.bol %diff %chat-update [%create (slav %p i.pax) pax]]
+  ?:
+    ?&
+      ?=(^ backlog-start)
+      (~(got by allow-history) pax)
+    ==
+    (paginate-messages pax u.box u.backlog-start)
```
Comment (Fang-) [inline pkg/arvo/app/chat-hook.hoon]: Why do we paginate instead of just sending whatever's requested? I don't think we fall over on big messages anymore.

File: pkg/arvo/sur/rw-security.hoon (hunk 4 above)
Comment (loganallenc) [reply pkg/arvo/sur/rw-security.hoon]: I think this is a small, positive change...

File: pkg/arvo/sur/chat-store.hoon (hunk 6 above)
Comment (loganallenc) [reply pkg/arvo/sur/chat-store.hoon]: I expect this to work just fine once all ships have updated given that my state import functions have been tested on ships with older state versions. As for this network message issue, seems like it'll only be a problem for ships on latest vs latest-1, which shouldn't last very long?

Comment (loganallenc) [reply pkg/arvo/sur/chat-store.hoon]: As for %message, I think it's semantically useful to have a "singular version". Idk, arguable.

File: pkg/arvo/app/chat-hook.hoon (hunk 7 above)
Comment (loganallenc) [reply pkg/arvo/app/chat-hook.hoon]: It scopes the migrate state function within this core to show that it's only used here.

File: pkg/arvo/app/chat-hook.hoon (hunk 8 above)
Comment (loganallenc) [reply pkg/arvo/app/chat-hook.hoon]: Idk... Big messages scare me.

File: pkg/arvo/app/chat-cli.hoon (hunk 2 above)
Comment (loganallenc) [reply pkg/arvo/app/chat-cli.hoon]: I know very little about this type of parsing, so I was cargo-culting. Thanks for your suggestions.

File: pkg/arvo/sur/rw-security.hoon (hunk 4 above)
Comment (Fang-) [reply pkg/arvo/sur/rw-security.hoon]: And I agree! Just saying it's bad practice to include it in a commit that does something completely unrelated. The change isn't at all detectable from the commit message, and it pollutes the diff of the change the commit is claiming to make, which is messier for review.

Not a big deal in this particular case, but gotta stay vigilant. (:

File: pkg/arvo/app/chat-hook.hoon (hunk 7 above)
Comment (Fang-) [reply pkg/arvo/app/chat-hook.hoon]: Oh, durr, I read this as a `^|` somehow. Nevermind!

**Comment by loganallenc** [comment]
I don't care about supporting the "negative index" case yet. I would prefer to keep the interface simple with a "request everything", "request to be updated with everything since I was last in the room", or "request only new messages, no old messages". This is a simpler api for the end user than providing them a combinatorial explosion of options about how much backlog to request.

File: pkg/arvo/app/chat-cli.hoon (hunk 1 above)
Comment (ohAitch) [reply pkg/arvo/app/chat-cli.hoon]: I think you want ++fuss from hoon.hoon

File: pkg/arvo/sur/chat-store.hoon (hunk 6 above)
Comment (Fang-) [reply pkg/arvo/sur/chat-store.hoon]: Yeah I don't doubt that app state will be fine.

> it'll only be a problem for ships on latest vs latest-1, which shouldn't last very long?

You think too highly of the current network. (^:

iirc the consequences were pretty harsh last time we messed this up. Big stack traces, unusable chats, etc. Worth testing and maybe even holding off until breach for.

---

> I think it's semantically useful to have a "singular version". Idk, arguable.

If it's semantics, then that difference should be explicit (instead of `%singular` and `%plural`). Implementation-wise, I've found it nice to stick to singular when you can, but just go all-in on plural when you can't. Having both is just interface bloat.

Comment (joemfb) [reply pkg/arvo/sur/chat-store.hoon]: This kind of change is supposed to work OTA -- if the message types become incompatible, the subscription is cleared (or the sender backs off, as appropriate). In practice, that process is incomplete and not particularly tuned, so it can be quite noisy.

A simple solution is to just use a different mark for backlog subscriptions. The receiver won't make the subscription if they don't have the mark, and the new type won't mess with current subscriptions. Then you can combine the types in the next breach.

File: pkg/arvo/sur/rw-security.hoon (hunk 4 above)
Comment (joemfb) [reply pkg/arvo/sur/rw-security.hoon]: Separate commits endorsed.

File: pkg/arvo/app/chat-hook.hoon (hunk 8 above)
Comment (joemfb) [reply pkg/arvo/app/chat-hook.hoon]: Pagination is generally righteous IMO, at least in cases where a client can practically make use of the individual pages. But there also aren't any problems with big messages that I'm aware of.

File: pkg/arvo/sur/chat-store.hoon (hunk 6 above)
Comment (loganallenc) [reply pkg/arvo/sur/chat-store.hoon]: This simple solution feels very messy to me... I'd much prefer noise in the dojo than making another mark.

Comment (joemfb) [reply pkg/arvo/sur/chat-store.hoon]: Tfw a developer is unwilling to make sacrifices for the experience of their users...

File: pkg/arvo/sur/rw-security.hoon (hunk 4 above)
Comment (loganallenc) [reply pkg/arvo/sur/rw-security.hoon]: I have now split this up into different commits. Laborious, but worth it!

Comment (joemfb) [reply pkg/arvo/sur/rw-security.hoon]: Positively virtuous!

File: pkg/arvo/app/chat-cli.hoon (hunk 3 above)
Comment (loganallenc) [reply pkg/arvo/app/chat-cli.hoon]: cool, will use +fall

Comment (loganallenc) [reply pkg/arvo/app/chat-cli.hoon]: done

File: pkg/arvo/app/chat-cli.hoon (hunk 1 above)
Comment (loganallenc) [reply pkg/arvo/app/chat-cli.hoon]: done, thanks!

Comment (Fang-) [reply pkg/arvo/app/chat-cli.hoon]: This uses `+fuss` now, but the code below doesn't, and both still want too many spaces.

Comment (loganallenc) [reply pkg/arvo/app/chat-cli.hoon]: I've tested out the ;create parser, and it only wants a single space in between each option. Currently doing ;join

Comment (Fang-) [reply pkg/arvo/app/chat-cli.hoon]: Oh huh, sorry, I think gh was showing me outdated code. (Or I was interpreting it as the old code.) This does indeed seem good. Carry on! (:

File: pkg/arvo/app/chat-cli.hoon (hunk 2 above)
Comment (loganallenc) [reply pkg/arvo/app/chat-cli.hoon]: this is done

**Comment by loganallenc** [comment]
Can I get another pass on this? @joemfb @philipcmonk @Fang-

File: pkg/interface/chat/src/js/components/lib/chat-input.js (hunk 5 above)
Comment (Fang-) [reply pkg/interface/chat/src/js/components/lib/chat-input.js]: Probably just remove these.

File: pkg/arvo/sur/chat-store.hoon (hunk 6 above)
Comment (Fang-) [reply pkg/arvo/sur/chat-store.hoon]: Seconding @joemfb's sentiment here. Whatever's messy about it we can clean up on next breach (that is, right now in a separate, breach-flagged PR). We're busy enough with support as-is, don't need more loud errors scaring people.

**Comment by Fang-** [comment]
Will do another code pass tomorrow!

File: pkg/arvo/sur/chat-store.hoon (hunk 6 above)
Comment (loganallenc) [reply pkg/arvo/sur/chat-store.hoon]: If you can't add a type to a tagged union OTA, then our support for OTAs doesn't exist imo.

Comment (vvisigoth) [reply pkg/arvo/sur/chat-store.hoon]: Call it what you will. I've seen the fall out of not doing this and it sucks and will mean emails that someone else has to respond to. Please do this and let's wrap this up.

File: pkg/arvo/app/chat-hook.hoon [hunk 9]
```diff
@@ -110,20 +131,23 @@
...
     (create-permission [%chat path.act] security.act)
   ::
       %add-synced
     ?>  (team:title our.bol src.bol)
-    =/  chat-path  [%mailbox (scot %p ship.act) path.act]
+    =/  chat-path=path  [%mailbox (scot %p ship.act) path.act]
     ?:  (~(has by synced) [(scot %p ship.act) path.act])
       [~ this]
     =.  synced  (~(put by synced) [(scot %p ship.act) path.act] ship.act)
+    =/  history=path  ?:(ask-history.act /0 /~)
```
Comment (Fang-) [inline pkg/arvo/app/chat-hook.hoon]: Wait, do we not want the empty path `/` here for the second case? Pretty sure it's more common to do `/some/resource` and then parse the absence of a trailing `/index`, instead of doing `/some/resource/nothing` explicitly. Certainly have never seen suffix `/~` before.

File: pkg/arvo/app/chat-hook.hoon (hunk 8 above)
Comment (Fang-) [reply pkg/arvo/app/chat-hook.hoon]: _Maybe_ this is unresolved since pagination here is a data-sending mechanism rather than part of the data-retrieving query. Since big messages aren't a problem, maybe just pull this logic out and send backlog all at once, to reduce complexity. But I don't feel super strongly.

File: pkg/arvo/app/chat-hook.hoon [hunk 10]
```diff
@@ -263,7 +341,7 @@
     ?.  =(u.shp src.bol)
       [~ this]
     :_  this(synced (~(del by synced) path.diff))
-    :-  (chat-poke diff)
+    :-  (chat-poke [%delete path.diff])
```
Comment (Fang-) [inline pkg/arvo/app/chat-hook.hoon]: Isn't this equivalent to the previous code (passing in `diff` as-is)?

Same for the `%message` case below. What am I missing?

Comment (loganallenc) [reply pkg/arvo/app/chat-hook.hoon]: Type nesting changes for `chat-base`, `chat-update`, and `chat-action`.

File: pkg/arvo/app/chat-cli.hoon [hunk 11]
```diff
@@ -39,12 +39,14 @@
...
-      [%create chat-security path (unit glyph)]     ::  create chat
+      ::
+      ::  create chat
+      [%create rw-security path (unit glyph) (unit ?)]
       [%delete path]                                ::  delete chat
       [%invite ?(%r %w %rw) path (set ship)]        ::  allow
       [%banish ?(%r %w %rw) path (set ship)]        ::  disallow
     ::
-      [%join target (unit glyph)]                   ::  join target
+      [%join target (unit glyph) (unit ?)]          ::  join target
```
[the remark this answers is not in the document]
Comment (loganallenc) [reply pkg/arvo/app/chat-cli.hoon]: In that case, I will keep it unit.

File: pkg/arvo/app/chat-hook.hoon [hunk 12]
```diff
@@ -33,42 +34,62 @@
...
+        synced.sta  synced
+        allow-history.sta  (create-allow-history synced)
+        invite-created  %.y
+    ==
+    this(+<+ sta)
+  ::
+  ++  create-allow-history
+    |=  synced=(map path ship)
+    ^-  (map path ?)
+    (~(run by synced) |=(shp=ship %.n))
```
[the remark this answers is not in the document]
Comment (loganallenc) [reply pkg/arvo/app/chat-hook.hoon]: thanks, will do.

File: pkg/interface/chat/src/js/components/lib/chat-input.js (hunk 5 above)
Comment (loganallenc) [reply pkg/interface/chat/src/js/components/lib/chat-input.js]: done

File: pkg/arvo/app/chat-store.hoon [hunk 13]
```diff
@@ -198,6 +199,35 @@
...
+  =*  envelope       i.envelopes.act
+  =*  letter         letter.envelope
+  =?  letter  &(?=(%code -.letter) ?=(~ output.letter))
+    =/  =hoon  (ream expression.letter)
+    letter(output (eval bol hoon))
+  =:  length.config.u.mailbox  +(length.config.u.mailbox)
+      number.envelope  +(length.config.u.mailbox)
+      envelopes.u.mailbox  (snoc envelopes.u.mailbox envelope)
+  ==
+  =.  inbox  (~(put by inbox) path.act u.mailbox)
```
[the remark this answers is not in the document]
Comment (philipcmonk) [reply pkg/arvo/app/chat-store.hoon]: Isn't the difference that this doesn't send the diffs individually?  You could factor out the duplicated code, though.

File: pkg/arvo/app/chat-hook.hoon [hunk 14]
```diff
@@ -157,14 +181,57 @@
...
+  ^-  (list move)
+  =/  moves=(list move)  ~
+  =/  end  (lent envelopes.mailbox)
+  ?:  |((gte start end) =(end 0))
+    moves
+  =.  envelopes.mailbox  (slag start `(list envelope)`envelopes.mailbox)
+  |-  ^-  (list move)
+  ?~  envelopes.mailbox
+    moves
+  ?:  (lte end 5.000)
```
Comment (philipcmonk) [inline pkg/arvo/app/chat-hook.hoon]: 5000 is a lot, have you tested that many to make sure it doesn't bail:meme?

File: pkg/arvo/app/chat-hook.hoon [hunk 15]
```diff
@@ -157,14 +181,57 @@
 ++  peer-mailbox
   |=  pax=path
   ^-  (quip move _this)
+  ?>  ?=(^ pax)
+  =/  last  (dec (lent pax))
+  =/  backlog-start=(unit @ud)
+    %+  rush
+      (snag last `(list @ta)`pax)
+    dem:ag
+  =>  .(pax `path`(oust [last last] `(list @ta)`pax))
```
[the remark this answers is not in the document]
Comment (loganallenc) [reply pkg/arvo/app/chat-hook.hoon]: yeah, TMI. I'll change to last 1.

File: pkg/arvo/app/chat-store.hoon (hunk 13 above)
[the remark this answers is not in the document]
Comment (loganallenc) [reply pkg/arvo/app/chat-store.hoon]: I've now factored out duplicated code.

File: pkg/arvo/app/chat-hook.hoon (hunk 9 above)
Comment (loganallenc) [reply pkg/arvo/app/chat-hook.hoon]: It's fine either way, but I actually like the /~ convention.

File: pkg/arvo/app/chat-hook.hoon (hunk 15 above)
[the remark this answers is not in the document]
Comment (loganallenc) [reply pkg/arvo/app/chat-hook.hoon]: done

File: pkg/arvo/app/chat-hook.hoon (hunk 12 above)
[the remark this answers is not in the document]
Comment (loganallenc) [reply pkg/arvo/app/chat-hook.hoon]: done

File: pkg/arvo/app/chat-hook.hoon (hunk 14 above)
Comment (loganallenc) [reply pkg/arvo/app/chat-hook.hoon]: Works just fine, just tested it.

File: pkg/arvo/sur/chat-store.hoon (hunk 6 above)
Comment (loganallenc) [reply pkg/arvo/sur/chat-store.hoon]: Working on this next.

Comment (loganallenc) [reply pkg/arvo/sur/chat-store.hoon]: I've tested it with the new "chat-two" marks, and it all works great.

**Comment by loganallenc** [comment]
Should be completely finished @Fang-.

File: pkg/arvo/lib/chat-json.hoon [hunk 16]
```diff
@@ -141,6 +134,23 @@
...
+::
+++  update-to-json
+  |=  upd=chat-update
+  =,  enjs:format
+  ^-  json
+  %+  frond  %chat-update
+  %-  pairs
+  :~
+    ?:  =(%message -.upd)
+      ?>  ?=(%message -.upd)
```
Comment (Fang-) [inline pkg/arvo/lib/chat-json.hoon]: Maybe I missed this before, but I'm curious: isn't just `?:  ?=(%message -.upd)` practically equivalent here? I don't recognize any TMI-sensitive situations in the code below.

Comment (loganallenc) [reply pkg/arvo/lib/chat-json.hoon]: Yes, it is equivalent. I'll make a note to go modify all of the JSON marks with that expression rather than the ?>  expression. (it's used across a variety of marks that i've written)

**Comment by loganallenc** [comment]
@jtobin This is ready to merge. Thanks!
